code/modules/food_and_drinks/kitchen_machinery/food_replicator.dm

```diff
@@ -0,0 +1,631 @@
+	var/list/reagent_tier4 = list("omnizine", "weak_omnizine", "sulfonal")
+
+	//memory_slots is the number of memory slots you can save dishes to for replication. Upgrading the scanning module improves this, allowing for more stored dish options.
+	var/memory_slots = 1
+
+	//Memory slot breakdown:
+	//"name" stores the dish's displayed name, forn ensuring replicated dishes also are named the same as the original (for renamed dishes mostly)
+	//"item" stores the food item path, for creating the replicated items of the correct type.
+	//"reagents" is a list of all the reagents in the food when it was scanned, though not all reagents may be replicated depending on the reagent_tier and reagents present.
+	var/list/memory1 = list("name" = null, "item" = null, "reagents" = list())
+	var/list/memory2 = list("name" = null, "item" = null, "reagents" = list())
+	var/list/memory3 = list("name" = null, "item" = null, "reagents" = list())
+	var/list/memory4 = list("name" = null, "item" = null, "reagents" = list())
+	var/obj/item/weapon/reagent_containers/food/snacks/loaded = null
+
+/obj/machinery/food_replicator/New()
+	..()
+	component_parts = list()
+	component_parts += new /obj/item/weapon/circuitboard/food_replicator(null)
+	component_parts += new /obj/item/weapon/stock_parts/micro_laser(null)
+	component_parts += new /obj/item/weapon/stock_parts/manipulator(null)
+	component_parts += new /obj/item/weapon/stock_parts/scanning_module(null)
+	component_parts += new /obj/item/weapon/stock_parts/matter_bin(null)
+	component_parts += new /obj/item/weapon/stock_parts/console_screen(null)
+	component_parts += new /obj/item/stack/cable_coil(null, 5)
+	RefreshParts()
+
+/obj/machinery/food_replicator/upgraded/New()
+	..()
+	component_parts = list()
+	component_parts += new /obj/item/weapon/circuitboard/food_replicator(null)
+	component_parts += new /obj/item/weapon/stock_parts/micro_laser/quadultra(null)
+	component_parts += new /obj/item/weapon/stock_parts/manipulator/femto(null)
+	component_parts += new /obj/item/weapon/stock_parts/scanning_module/triphasic(null)
+	component_parts += new /obj/item/weapon/stock_parts/matter_bin/bluespace(null)
+	component_parts += new /obj/item/weapon/stock_parts/console_screen(null)
+	component_parts += new /obj/item/stack/cable_coil(null, 5)
+	RefreshParts()
+
+/obj/machinery/food_replicator/Destroy()
+	clear_memory(-1)		//just in case, probably don't need to do this.
+	eject_dish()
+	if(food_paste)			//if we have food paste left, let's just dump a pile of goop on the floor to represent it being wasted
+		if(food_paste >= 100)	//a lot of food paste, better make it a big pile of goop
+			new /obj/effect/decal/cleanable/molten_object/large(get_turf(src))
+		else
+			new /obj/effect/decal/cleanable/molten_object(get_turf(src))
```

Comment:
eeeeeggh something that leaves residue behind on `Destroy`
Won't this leave behind a mess if you crowbar this apart